### DISCUSSION TITLE
external midi to modulator no midi output

### DISCUSSION THREAD

**Discussion by thethirdapple** [RomanKubiak/ctrlr#540] [Q&A] [answered]
Greetings,

With an external midi controller I am able to input midiCC which adjusts a rotary slider, but it does not output midi data.
When same modulator is adjusted with cursor it outputs midi data...

A perplexed new Ctrlr user?!?! 
Is this a LUA script or a setting directly thru the side menu?

If someone could point me in the correct direction or to "complete" documentation, I have searched GitHub forum, og Ctrlr forum, google machine... finding scraps of really good info.

Ctrlr: Version = 5.3.198, Build date = Tue Mar 15 15:53:07 PDT 2016, Branch = Nightly, Juce = 4.0.2, libusb = 1.0.19, liblo = 0.28, 
running on: Mac OS 10.13.6

**Comment by damiensellier** [comment]
EDIT: I wrote a previous post, but I got the same results 
The message sent from the controller is not sent "Thru" to avoid loops etc

https://ctrlr.org/forums/topic/midi-device-settings-controllerin-device/

You need a lua script sniffing the input and sending it to your output.

**Reply by thethirdapple** [reply]
Appreciate the link... I had seen that post on the old forum and did find a work around for what I'm look to do but have not been able to replicate it. Not even stoned at the moment!

Essentially looking to translate incoming cc to outgoing sysex for the Korg Karma...
had a working set up of two sliders:  one with incoming cc from BCR (a whatever plastic crap knob controller) and a second with variable sysex outgoing. Had them linked through the "link to modulator" option somehow and worked once but unable replicate after Ctrlr crash. The 1st slide was "adjusting" the second slider which output the corresponding sysex.

I guess bugs can also work sometimes in our favor.

Will go thru a host and try it out as per ATOM. I usually use software called Bidule from Plogue, which is a great host for everything including Ctrlr.

![Screen Shot 2023-02-17 at 10 12 38 PM](https://user-images.githubusercontent.com/125695472/219828777-7fd145f9-e905-49c5-8693-7614dc294fe1.png)

**Reply by damiensellier** [reply]
I edited my message below, it's working now.

**Reply by damiensellier** [reply]
And if you need to convert CC to sysex you'll need a LUA script for the modulator "when the modulator value change". 
It will get the modulator value (set from your BCR as a CC), and needs a translation to sysex.
You'll need to share with us the target sysex formula message from the manual so that we could help ;) 

Cheers
Damien

**Reply by thethirdapple** [reply]
Thank you for your help with this. Indeed if a scripting we shall go, lets go! It is getting late here so I will pick this up in the morning. Peace 

the BCR is sending on channel 1 -  CC#9 (could be any channel or CC#) - value 1-4 
the Karma is expecting Sysex F0 42 30 5D 41 03 00 00 00 09 00 xx F7 

the "xx" is the variable 01, 02, 03, 04 which is picked up from the value 1-4 of the slider, which works fine as an independent slider, when set to send sysex with that "xx" formula.

This sysex controls the type of tuning on the Korg Karma between 01)Equal Temperament, 02)Werckmeister, 03)Pure Major, 04)Pure Minor...

**Comment by damiensellier** [comment]
Alright so you have to create a "midi thru" from the panel preferences by calling a script "**when a pannel receives a midi message**:

CLICK THE CANVAS and OPEN THE PANEL MODE cmd+E, it's in the middle:
![CLICK THE CANVAS and OPEN THE PANEL MODE cmd+E](https://user-images.githubusercontent.com/55164385/219828935-ffe19cf6-3949-45d0-b4a2-7f8052770e55.png)


the script is simple : 
```

--
-- Called when a panel receives a midi message (does not need to match any modulator mask)
-- @midi   CtrlrMidiMessage object
--

midiThru = function(--[[ CtrlrMidiMessage --]] midi)

	panel:sendMidiMessageNow(midi) -- Send THRU Midi Message

end
```

This way every message coming in will be repeated to the output.



For the modulator, you just need to tell what CC it is etc as usual: 

Click the modulator and in the middle of the pannel you'll see the message type details:

![Capture d’écran, le 2023-02-17 à 22 15 36](https://user-images.githubusercontent.com/55164385/219828966-8c28dc39-388c-424b-a650-29271132c1be.png)

And in the MIDI menu, you need to select both "input" and "controller" in the MIDI menu to the same device with the same channel.

INPUT
![INPUT](https://user-images.githubusercontent.com/55164385/219829004-3c1cb045-e3ec-4e40-8e36-c0a1077ffbc7.png)

CONTROLLER
![CONTROLLER](https://user-images.githubusercontent.com/55164385/219829019-df73047d-14f0-4b34-a2a5-66e571736855.png)

OUTPUT
![OUTPUT](https://user-images.githubusercontent.com/55164385/219829022-cef8741a-c209-417b-83dc-8ac7a2fdb417.png)


I attached the sample panel 

It's working this way, both the modulator on click and the controller device send the message to the selected output.


[midiThru From Controller Input_1_0_2023-02-17_22-09.panel.zip](https://github.com/RomanKubiak/ctrlr/files/10772974/midiThru.From.Controller.Input_1_0_2023-02-17_22-09.panel.zip)

**Reply by thethirdapple** [reply]
The standalone build is buggy on my machine, seems more stable as a plug in for sure and I appreciate the script, but is this not simply the same as using the Ctrlr drop down midi thru menu?

So yes, both the controller and the slider are putting out the same CC info.

The goal is not a straight "thru" but a Ctrlr interpret (transpose or transform) the CC as something else. Such that in this simple proof of concept have Ctrlr add a value to the CC or an all together new Sysex  output based on a CC input.

**Reply by damiensellier** [reply]
oh boy, yeah probably the same as the midi thru option, I never used it so I didn't even know it existed lol 

you need a LUA script to translate your incoming midi message when the modulator value change. 
Can you please post the output message you're looking for? It's probably written in your device manual or midi sysex implementation documentation.

I'll help you with it ;)

**Reply by thethirdapple** [reply]
replied to your earlier reply with the sysex formula... I truly appreciate your help although just remember: if you give a mouse a cookie...
peace

**Reply by damiensellier** [reply]
haha nevermind I'll open the door so that you understand the concept. It's basically based on:


```
ccToSysex = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	local modValue = mod:getModulatorValue() -- to get the current value of your modulator
	panel:sendMidiMessageNow(CtrlrMidiMessage(MemoryBlock({0xF0, 0x42, 0x30, 0x5D, 0x41, 0x03, 0x00, 0x00, 0x00, 0x00, 0x09, 0x00, modValue, 0xF7}))) -- F0 42 30 5D 41 03 00 00 00 09 00 xx F7

end
```

We need to filter the input message to check if it's the right CC and also setup the 4 names to each value 0,1,2,3

I'll post the full code later, stay tuned

**Comment by damiensellier** [comment]
@thethirdapple  There we go! Don't forget my cookie ...

[BCR CC TO KORG SYSEX](https://github.com/RomanKubiak/ctrlr/files/10773496/KORG.SYSEX.FROM.CC_1_0_2023-02-18_01-26.panel.zip)

![PREVIEW](https://user-images.githubusercontent.com/55164385/219845407-a4dfb9bd-be2c-4b84-b401-8941eb83d7cb.gif)

[BCR CC TO KORG SYSEX](https://github.com/RomanKubiak/ctrlr/files/10773496/KORG.SYSEX.FROM.CC_1_0_2023-02-18_01-26.panel.zip)

You just need to set your own midi inputs and outputs. I let you check all the settings for the modulator. I muted the midi output from the modulator operator and superseeded it with the lua script sending sysex. On my setup it was sending sysex from both the mouse control and the midi controller device.

Damien

**Comment by dnaldoog** [comment, marked as answer]
You probably need to do this in lua.

For a dynamic solution you would need to have some way for the user to change cc mapping, but for a static solution here is a suggestion:

<pre>
--
-- Called when a panel receives a midi message (does not need to match any modulator mask)
-- @midi   CtrlrMidiMessage object
--

myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)
        panel:getModulatorByName(mapping[number]):setModulatorValue(value, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) 
	--[[ 
		fallback value generated in case a cc number comes in but 
		there is no corresponding control mapped to that number
		]]
    t[k] = "nullbin"
    utils.warnWindow("Error", "number not mapped to a control") -- comment this out to die silently
    return t[k]
end

mapping = {
    "resonance",
    "cutoff"
}
setmetatable(mapping, meta)
</pre>

Here you would add the name of each control on the panel to the table `mapping`. If a cc number comes in but is not mapped then the value is sent to a hidden uiSlider like a /dev/null kind of thing.

Uncomment the annoying message for the script to return silently.

See attached:

[map cc to sysex_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/10773705/map.cc.to.sysex_1_0.zip)

----

![ccmapp10](https://user-images.githubusercontent.com/76831683/219851004-c4eea0aa-dd63-4f0c-9395-1d0ff02c3cc5.png)

----

P.S. With lua tables you can also do things like:

<pre>
mapping = {
    [23]="resonance",
    [30]="cutoff"
}
</pre>

so you can specify the exact number you want to assign without having a list of 30 table values just to get to `30` in this case.

**Reply by damiensellier** [reply]
Hi John, happy new year, hope everything is alright on your side!

I didn't know about : 
`midi:getType() == 0 then -- it is a CC message`
That's really helpful, I'll use it on my pannels because previously I was doing something complicated to filter the type of messages: 
```
midiType = midi:getData():getByte(0) 
if midiType >= 0xB0 and midiType <= 0xBF then ....
```

We learn things everyday :) 
And your method is dynamic and easy to work with, mine is static and requires a script for each slider, so if the OP wants to deploy his panel for all the parameters, you did it right!
Have a nice day
Damien

**Reply by dnaldoog** [reply]
Hi Damien - Happy New Year to you too! Pretty cold here but probably nothing like what you are used to no doubt!

Yes I used to do something similar to you with CC messages, but I knew that `if midi:getType() == 5` is for sysex, so I was able to find `0` is for CC on the [old forum](https://ctrlr.org/forums/topic/uicombo-not-responding-to-midi-cc-3/page/2/#post-84435) (only just now), so eventually I will change old panels too when I get around to it!

But actually this new method I posted is also much simpler than code I have written before for ccmapping, so you do certainly live and learn and sometimes answering questions leads to new insights.

Another way you could do what you used to do, because speed is important here could be:

`if bit.band(midiType,240)==176 then ...` but it's not necessary now!

JG

**Reply by thethirdapple** [reply]
Greetings,

FANTASTIC, will fire up the machine after a few more shots of coffee and implement this…

The dynamic attribution is for sure the way I would like to go, the KARMA has 407 realtime parameters per “module” and there are 4 modules… will need a few more BCR’s for knob fiddling ;) 

Orbit Cookies for everyone!!!

peace

**Comment by dnaldoog** [comment]
If that panel is working, there is another modification that is worth doing.
Please read here for an explanation https://github.com/RomanKubiak/ctrlr/discussions/482#discussion-4355575

`panel:getModulatorByName()` is inefficient so this version 2.0 uses a caching system, which will be important in cases like this. 

New code:

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)
        _m(mapping[number]):setModulatorValue(value, false, true, false)
    end
end
</pre>

This panel uses a memoized table reference to each ctrlr component - it should make a difference once you get the 407 parameters all set up!

[map cc to sysex_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/10774363/map.cc.to.sysex_2_0.zip)

**Reply by thethirdapple** [reply]
NICE! Thank you for the preemptive responsiveness… this is a great community!

going to take it slowly to make sure I understand whats happening and why… my audio and midi patchbays stare at me everyday, naggingly reminding me to label and map connections.

Have not done any “serious” coding for nearly 10 years. And way way back (90’s) in the day was a hardcore Quicktime developer,  then did a bunch of VRML which lead to scripting Lightwave 3D environments in the early 2000’s.

Was initially brought to Ctrlr for the Tg33 panel a while ago (i refuse to use midiquest). Trying to get away from a screen when jamming. I have a bunch of older synth’s and equipment which predate midi, but I insist on having the entire studio communicating all together all the time…

One controller to rule them all…

peace

**Comment by thethirdapple** [comment]
Greetings,

First of all: THANK YOU to dnaldoog, damiensellier for your awesome assistance with this proof. And obviously to all whom have created and contributed to this platform and forum...

The mapping function works like a charm, overall the scripts are functioning perfectly. The ability to copy paste components with minimal recoding is going to be a dream considering the 407 realtime KARMA parameters to be fiddled with.

I did go down the image resource rabbit hole today and finally figured out to simply run two versions of Ctrlr, 5.3.85 to import and manage images, save, then reopen panel in 5.3.198 to build. 

Now begins the actual work of deciding just how in-depth I go with this panel so as to share something useful...

peace

![proofof concept_v3_2 2](https://user-images.githubusercontent.com/125695472/219903649-2b94d585-8f62-4721-9170-cd48803f5b7e.png)

**Reply by dnaldoog** [reply]
That’s great to hear! 🥳

**Reply by damiensellier** [reply]
Congrats mate, welcome to the Ctrlr family!
